src/osp/Active/ActiveScene.h

```diff
@@ -332,4 +344,47 @@ struct ACompCamera
 };
 
 
+template<typename FUNC_T>
+void ActiveScene::hierarchy_traverse(ActiveEnt root, FUNC_T callable)
+{
+    using osp::active::ACompHierarchy;
+
+    std::vector<ActiveEnt> parentNextSibling;
+    ActiveEnt currentEnt = root;
+
+    parentNextSibling.reserve(16);
+    while (true)
+    {
+        ACompHierarchy &hier = reg_get<ACompHierarchy>(currentEnt);
+
+        if (!callable(currentEnt)) { return; }
+
+        if (hier.m_childCount)
+        {
+            // entity has some children
+            currentEnt = hier.m_childFirst;
+
+
+            // save next sibling for later if it exists
+            if (hier.m_siblingNext != entt::null)
+            {
+                parentNextSibling.push_back(hier.m_siblingNext);
```

Comment:
Would it make sense to store an std::reference_wrapper to the ACompHierarchy into the vector instead of ActiveEnt?

Reply:
It should be possible, but it would only rearrange the entire function without much improvements. `reg_get` would still have to be called for every entity iterated, which isn't expensive at all. References to components can be invalidated if new components are added.